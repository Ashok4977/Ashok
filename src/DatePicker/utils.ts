import deepmerge from 'deepmerge'
import {DatePickerConfiguration, isMultiSelection, isRangeSelection, Selection} from './types'
import {defaultConfiguration} from './DatePickerProvider'

export const sanitizeDate = (date: Date | string) => new Date(new Date(date).toDateString())

export const getInitialFocusDate = (selection?: Selection) => {
  if (!selection) return sanitizeDate(new Date())

  if (selection instanceof Date) {
    return sanitizeDate(selection)
  } else if (isMultiSelection(selection)) {
    return sanitizeDate(selection[0])
  } else if (isRangeSelection(selection)) {
    return sanitizeDate(selection.from)
  } else {
    return sanitizeDate(new Date())
  }
}

export const initializeConfigurations = (externalConfig?: DatePickerConfiguration) => {
  let sanitizedConfig = {}
  if (externalConfig) {
    sanitizedConfig = Object.entries(externalConfig).filter(item => item[1] !== undefined && item[1] !== null)
    for (const item of sanitizedConfig) {
      if (item instanceof Date) {
        sanitizeDate(item)
      }
    }
  }
  return deepmerge(defaultConfiguration, sanitizedConfig)
}
