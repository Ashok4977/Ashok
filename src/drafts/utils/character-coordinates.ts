// Modified from https://github.com/koddsson/textarea-caret-position, which was
// itself forked from https://github.com/component/textarea-caret-position.

// Note that some browsers, such as Firefox, do not concatenate properties
// into their shorthand (e.g. padding-top, padding-bottom etc. -> padding),
// so we have to list every single property explicitly.
const propertiesToCopy = [
  'direction', // RTL support
  'boxSizing',
  'width', // on Chrome and IE, exclude the scrollbar, so the mirror div wraps exactly as the textarea does
  'height',
  'overflowX',
  'overflowY', // copy the scrollbar for IE

  'borderTopWidth',
  'borderRightWidth',
  'borderBottomWidth',
  'borderLeftWidth',
  'borderStyle',

  'paddingTop',
  'paddingRight',
  'paddingBottom',
  'paddingLeft',

  // https://developer.mozilla.org/en-US/docs/Web/CSS/font
  'fontStyle',
  'fontVariant',
  'fontWeight',
  'fontStretch',
  'fontSize',
  'fontSizeAdjust',
  'lineHeight',
  'fontFamily',

  'textAlign',
  'textTransform',
  'textIndent',
  'textDecoration', // might not make a difference, but better be safe

  'letterSpacing',
  'wordSpacing',

  'tabSize',
  'MozTabSize' as 'tabSize'
] as const

function getCaretCoordinates(
  element: HTMLTextAreaElement | HTMLInputElement,
  position: number,
  options?: {debug: boolean}
) {
  const debug = (options && options.debug) || false
  if (debug) {
    const el = document.querySelector('#input-textarea-caret-position-mirror-div')
    if (el) el.parentNode?.removeChild(el)
  }

  const isFirefox = 'mozInnerScreenX' in window

  // The mirror div will replicate the textarea's style
  const div = document.createElement('div')
  div.id = 'input-textarea-caret-position-mirror-div'
  document.body.appendChild(div)

  const style = div.style
  const computed = window.getComputedStyle(element)
  const isInput = element.nodeName === 'INPUT'

  // Default textarea styles
  style.whiteSpace = 'pre-wrap'
  if (!isInput) style.wordWrap = 'break-word' // only for textarea-s

  // Position off-screen
  style.position = 'absolute' // required to return coordinates properly
  if (!debug) style.visibility = 'hidden' // not 'display: none' because we want rendering

  // Transfer the element's properties to the div
  for (const prop of propertiesToCopy) {
    if (isInput && prop === 'lineHeight') {
      // Special case for <input>s because text is rendered centered and line height may be != height
      if (computed.boxSizing === 'border-box') {
        const height = parseInt(computed.height)
        const outerHeight =
          parseInt(computed.paddingTop) +
          parseInt(computed.paddingBottom) +
          parseInt(computed.borderTopWidth) +
          parseInt(computed.borderBottomWidth)
        const targetHeight = outerHeight + parseInt(computed.lineHeight)
        if (height > targetHeight) {
          style.lineHeight = `${height - outerHeight}px`
        } else if (height === targetHeight) {
          style.lineHeight = computed.lineHeight
        } else {
          style.lineHeight = '0'
        }
      } else {
        style.lineHeight = computed.height
      }
    } else if (!isInput && prop === 'width' && computed.boxSizing === 'border-box') {
      // With box-sizing: border-box we need to offset the size slightly inwards.  This small difference can compound
      // greatly in long textareas with lots of wrapping, leading to very innacurate results if not accounted for.
      // Firefox will return computed styles in floats, like `0.9px`, while chromium might return `1px` for the same element.
      // Either way we use `parseFloat` to turn `0.9px` into `0.9` and `1px` into `1`
      const totalBorderWidth = parseFloat(computed.borderLeftWidth) + parseFloat(computed.borderRightWidth)
      // When a vertical scrollbar is present it shrinks the content. We need to account for this by using clientWidth
      // instead of width in everything but Firefox. When we do that we also have to account for the border width.
      const width = isFirefox ? parseFloat(computed[prop]) - totalBorderWidth : element.clientWidth + totalBorderWidth
      style[prop] = `${width}px`
    } else {
      style[prop] = computed[prop]
    }
  }

  if (isFirefox) {
    // Firefox lies about the overflow property for textareas: https://bugzilla.mozilla.org/show_bug.cgi?id=984275
    if (element.scrollHeight > parseInt(computed.height)) style.overflowY = 'scroll'
  } else {
    style.overflow = 'hidden' // for Chrome to not render a scrollbar; IE keeps overflowY = 'scroll'
  }

  div.textContent = element.value.substring(0, position)
  // The second special handling for input type="text" vs textarea:
  // spaces need to be replaced with non-breaking spaces - http://stackoverflow.com/a/13402035/1269037
  if (isInput) div.textContent = div.textContent.replace(/\s/g, '\u00a0')

  const span = document.createElement('span')
  // Wrapping must be replicated *exactly*, including when a long word gets
  // onto the next line, with whitespace at the end of the line before (#7).
  // The  *only* reliable way to do that is to copy the *entire* rest of the
  // textarea's content into the <span> created at the caret position.
  // For inputs, just '.' would be enough, but no need to bother.
  span.textContent = element.value.substring(position) || '.' // || because a completely empty faux span doesn't render at all
  div.appendChild(span)

  const coordinates = {
    top: span.offsetTop + parseInt(computed['borderTopWidth']),
    left: span.offsetLeft + parseInt(computed['borderLeftWidth']),
    height: parseInt(computed['lineHeight'])
  }

  if (debug) {
    span.style.backgroundColor = '#aaa'
  } else {
    document.body.removeChild(div)
  }

  return coordinates
}

export type Coordinates = {
  top: number
  left: number
}

/**
 * Obtain the coordinates (px) of the bottom left of a character in an input, relative to the
 * top-left corner of the input itself.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 * @param adjustForScroll Control whether the returned value is adjusted based on scroll position.
 */
export const getCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number,
  adjustForScroll = true
): Coordinates => {
  if (!input) return {top: 0, left: 0}

  // word-wrap:break-word breaks the getCaretCoordinates calculations (a bug), and word-wrap has
  // no effect on input element anyway
  if (input instanceof HTMLInputElement) input.style.wordWrap = ''

  let coords = getCaretCoordinates(input, index)

  // The library calls parseInt on the computed line-height of the element, failing to account for
  // the possibility of it being 'normal' (another bug). In that case, fall back to a rough guess
  // of 1.2 based on MDN: "Desktop browsers use a default value of roughly 1.2".
  if (isNaN(coords.height)) coords.height = parseInt(getComputedStyle(input).fontSize) * 1.2

  // Sometimes top is negative, incorrectly, because of the wierd line-height calculations around
  // border-box sized single-line inputs.
  coords.top = Math.abs(coords.top)

  // For some single-line inputs, the rightmost character can be accidentally wrapped even with the
  // wordWrap fix above. If this happens, go back to the last usable index
  let adjustedIndex = index
  while (input instanceof HTMLInputElement && coords.top > coords.height) {
    coords = getCaretCoordinates(input, --adjustedIndex)
  }

  const scrollTopOffset = adjustForScroll ? -input.scrollTop : 0
  const scrollLeftOffset = adjustForScroll ? -input.scrollLeft : 0

  return {top: coords.top + coords.height + scrollTopOffset, left: coords.left + scrollLeftOffset}
}

/**
 * Obtain the coordinates of the bottom left of a character in an input relative to the top-left
 * of the page.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 */
export const getAbsoluteCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number
): Coordinates => {
  const {top: relativeTop, left: relativeLeft} = getCharacterCoordinates(input, index, true)
  const {top: viewportOffsetTop, left: viewportOffsetLeft} = input?.getBoundingClientRect() ?? {top: 0, left: 0}

  return {
    top: viewportOffsetTop + relativeTop,
    left: viewportOffsetLeft + relativeLeft
  }
}
