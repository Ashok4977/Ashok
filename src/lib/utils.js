
/**
 * Splice items from an Array for with the test function returns
 * truthy, modifying the original list and returning an array of
 * the spliced items.
 *
 * ```js
 * const children = React.Children.toArray(props.children)
 * const buttons = splice(children, child => child.type === Button)
 * ```
 */
const splice = (list, test) => {
  return list.reduce((acc, item, i) => {
    if (test(item)) {
      return acc.concat(list.splice(i, 1))
    }
    return acc
  }, [])
}

export {splice}
