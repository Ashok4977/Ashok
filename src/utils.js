import chroma from 'chroma-js'
import {css} from 'styled-components'
import {theme} from 'styled-system'

// color shortcuts
export const color = name => theme(`colors.${name}`)
export const black = color('black')
export const white = color('white')

/**
 * Reduce a value to a primitive as long as it is a function.
 *
 * interpolate(() => 'x') // returns 'x'
 * interpolate(() => () => 'y') // returns 'y'
 */
export const interpolate = (value, context, props) => {
  while (typeof value === 'function') {
    value = value.call(context, props)
  }
  return value
}

/**
 * This decorator interpolates all of the function's arguments by calling any
 * functions in the provided context and with the provided arguments. In other
 * words, the function you provide is guaranteed to have all of its arguments
 * "called" if they are functions. Example:
 *
 * const rgba = interpolator((rgb, a) => chroma(rgb).alpha(a).css())
 * const fadeed = rgba(theme('colors.black'), .5) // returns a function!
 * faded({}) // probably throws an error because props.theme isn't set
 */
export const interpolator = f => {
  return (...outer) => {
    return function(props) {
      return f(...[...outer].map(a => interpolate(a, this, props)))
    }
  }
}

export const rgba = interpolator(       (rgb, alpha)    => chroma(rgb).alpha(alpha).css())
export const lighten = interpolator(    (value, amount) => chroma(value).brighten(amount).css())
export const darken = interpolator(     (value, amount) => chroma(value).darken(amount).css())
export const desaturate = interpolator( (value, amount) => chroma(value).desaturate(amount).css())
export const mix = interpolator(        (a, b, amount)  => chroma.mix(a, b, amount).css())
