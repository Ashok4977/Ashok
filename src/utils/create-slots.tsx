import React from 'react'
import useLayoutEffect from './useIsomorphicLayoutEffect'

/** createSlots is a factory that can create a
 *  typesafe Slots + Slot pair to use in a component definition
 *  For example: ActionList.Item uses createSlots to get a Slots wrapper
 *  + Slot component that is used by LeadingVisual, Description
 */
const createSlots = <SlotNames extends string>(_slotNames: SlotNames[]) => {
  type Slots = {
    [key in SlotNames]?: React.ReactNode
  }

  type ContextProps = {
    registerSlot: (name: SlotNames, contents: React.ReactNode) => void
    unregisterSlot: (name: SlotNames) => void
  }
  const SlotsContext = React.createContext<ContextProps>({
    registerSlot: () => null,
    unregisterSlot: () => null
  })

  const Slots: React.FC<{
    children: (slots: Slots) => React.ReactNode
  }> = ({children}) => {
    const [slots, setSlots] = React.useState<Slots>({})

    const registerSlot = React.useCallback(
      (name: SlotNames, contents: React.ReactNode) => setSlots(prevSlots => ({...prevSlots, [name]: contents})),
      []
    )

    // Slot can be removed from the tree as well,
    // we need to unregister them from the slot
    const unregisterSlot = React.useCallback(
      (name: SlotNames) =>
        setSlots(prevSlots => {
          const newSlots = {...prevSlots}
          delete newSlots[name]
          return newSlots
        }),
      []
    )

    return <SlotsContext.Provider value={{registerSlot, unregisterSlot}}>{children(slots)}</SlotsContext.Provider>
  }

  const Slot: React.FC<{
    name: SlotNames
    children: React.ReactNode
  }> = ({name, children}) => {
    const {registerSlot, unregisterSlot} = React.useContext(SlotsContext)

    useLayoutEffect(() => {
      registerSlot(name, children)
      return () => unregisterSlot(name)
    }, [name, children, registerSlot, unregisterSlot])

    return null
  }

  return {Slots, Slot}
}

export default createSlots
