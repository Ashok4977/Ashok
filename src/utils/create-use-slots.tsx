import React from 'react'
import {useForceUpdate} from './use-force-update'

/** Create useSlots is a factory that can create a typed
 *  useSlots + Slot component combination for a component
 *  For example: ActionList uses createUseSlots to get a
 *  useSlot hook that can be used inside each Item
 */
export const createUseSlots = <SlotName extends string>(slotNames: SlotName[]) => {
  type ContextType = {
    registerSlot: (name: SlotName, contents: React.ReactNode) => void
    unregisterSlot: (name: SlotName) => void
  }
  const SlotsContext = React.createContext<ContextType>({registerSlot: () => null, unregisterSlot: () => null})

  const useSlots = () => {
    // Double render strategy
    // when the effect is run for the first time,
    // all the children have mounted = registed themself in slot.
    // we re-render the Item component to re-render with filled slots.
    const rerenderWithSlots = useForceUpdate()
    const [isMounted, setIsMounted] = React.useState(false)

    const slotsDefinition: {[key in SlotName]?: JSX.Element | null} = {}
    slotNames.map(name => {
      slotsDefinition[name] = null
    })
    const slotsRef = React.useRef<{[key in SlotName]?: React.ReactNode}>(slotsDefinition)

    // fires once after all the slots are registered
    React.useLayoutEffect(() => {
      setIsMounted(true)
      rerenderWithSlots()
    }, [rerenderWithSlots])

    const registerSlot = React.useCallback(
      (name: SlotName, contents: React.ReactNode) => {
        slotsRef.current[name] = contents

        // if something has changed?

        // don't render until the component mounts = all slots are registered
        if (isMounted) rerenderWithSlots()
      },
      [isMounted, rerenderWithSlots]
    )

    // Item.* can be removed from the DOM,
    // we need to unregister them from the slot
    const unregisterSlot = React.useCallback(
      (name: SlotName) => {
        slotsRef.current[name] = null
        rerenderWithSlots()
      },
      [rerenderWithSlots]
    )

    const SlotsProvider: React.FC = ({children}) => {
      return <SlotsContext.Provider value={{registerSlot, unregisterSlot}}>{children}</SlotsContext.Provider>
    }

    return {slots: slotsRef.current, SlotsProvider}
  }

  const Slot: React.FC<{name: SlotName}> = ({name, children}) => {
    const {registerSlot, unregisterSlot} = React.useContext(SlotsContext)

    React.useLayoutEffect(() => {
      registerSlot(name, children)
      return () => unregisterSlot(name)
      // registerSlot and unregisterSlot are created by the ItemContext,
      // we can safely ignore them because they will not change between renders
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [name, children])

    return null
  }

  return {useSlots, Slot}
}
