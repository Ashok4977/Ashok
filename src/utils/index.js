
export function defined(val) {
  return val !== null && typeof val !== 'undefined'
}

export function definedFallback(input, fallback) {
  return defined(input) ? input : fallback
}
