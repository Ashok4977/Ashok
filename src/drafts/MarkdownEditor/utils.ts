import {isMacOS} from '@primer/behaviors/utils'

export type InputState = {
  readonly value: string
  readonly selectionStart: number
  readonly selectionEnd: number
}

export const getSelectedLineRange = (inputState: InputState): [number, number] => {
  // Subtract one from the caret position so the newline found is not the one _at_ the caret position
  // then add one because we don't want to include the found newline. Also changes -1 (not found) result to 0
  const start = inputState.value.lastIndexOf('\n', inputState.selectionStart - 1) + 1

  // activeLineEnd will be the index of the next newline inclusive, which works because slice is last-index exclusive
  let end = inputState.value.indexOf('\n', inputState.selectionEnd)
  if (end === -1) end = inputState.value.length

  return [start, end]
}

export const markdownComment = (text: string) => `<!-- ${text.replaceAll('--', '\\-\\-')} -->`

export const markdownLink = (text: string, url: string) =>
  `[${text.replaceAll('[', '\\[').replaceAll(']', '\\]')}](${url.replaceAll('(', '\\(').replaceAll(')', '\\)')})`

export const markdownImage = (altText: string, url: string) => `!${markdownLink(altText, url)}`

export const isModifierKey = (event: KeyboardEvent | React.KeyboardEvent<unknown>) =>
  isMacOS() ? event.metaKey : event.ctrlKey

export function isMultipleLines(string: string): boolean {
  return string.trim().split('\n').length > 1
}
