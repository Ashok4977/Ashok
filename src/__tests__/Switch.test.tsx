import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import {render} from '@testing-library/react'
import {Switch} from '..'
import {behavesAsComponent, checkExports, checkStoriesForAxeViolations} from '../utils/testing'
import userEvent from '@testing-library/user-event'

const SWITCH_LABEL_TEXT = 'Switch label'

behavesAsComponent({
  Component: Switch,
  options: {skipAs: true}
})
checkExports('Switch', {
  default: Switch
})
it('renders a switch that is turned off', () => {
  const {getByLabelText} = render(
    <>
      <div id="switchLabel">{SWITCH_LABEL_TEXT}</div>
      <Switch aria-labelledby="switchLabel" />
    </>
  )
  const toggleSwitch = getByLabelText(SWITCH_LABEL_TEXT)

  expect(toggleSwitch).toHaveAttribute('aria-checked', 'false')
})
it('renders a switch that is turned on', () => {
  const {getByLabelText} = render(
    <>
      <div id="switchLabel">{SWITCH_LABEL_TEXT}</div>
      <Switch aria-labelledby="switchLabel" defaultOn />
    </>
  )
  const toggleSwitch = getByLabelText(SWITCH_LABEL_TEXT)

  expect(toggleSwitch).toHaveAttribute('aria-checked', 'true')
})
it('renders a switch that is disabled', () => {
  const {getByLabelText} = render(
    <>
      <div id="switchLabel">{SWITCH_LABEL_TEXT}</div>
      <Switch aria-labelledby="switchLabel" disabled />
    </>
  )
  const toggleSwitch = getByLabelText(SWITCH_LABEL_TEXT)

  expect(toggleSwitch).toHaveAttribute('aria-disabled', 'true')
  expect(toggleSwitch).toHaveAttribute('aria-checked', 'false')
  userEvent.click(toggleSwitch)
  expect(toggleSwitch).toHaveAttribute('aria-checked', 'false')
})
it("renders a switch who's state is loading", () => {
  const {getByLabelText, container} = render(
    <>
      <div id="switchLabel">{SWITCH_LABEL_TEXT}</div>
      <Switch aria-labelledby="switchLabel" isLoading />
    </>
  )
  const toggleSwitch = getByLabelText(SWITCH_LABEL_TEXT)
  const loadingSpinner = container.querySelector('svg')

  expect(loadingSpinner).toBeDefined()
  expect(toggleSwitch).toHaveAttribute('aria-disabled', 'true')
  expect(toggleSwitch).toHaveAttribute('aria-checked', 'false')
  userEvent.click(toggleSwitch)
  expect(toggleSwitch).toHaveAttribute('aria-checked', 'false')
})
it('switches from off to on uncontrolled', () => {
  const {getByLabelText} = render(
    <>
      <div id="switchLabel">{SWITCH_LABEL_TEXT}</div>
      <Switch aria-labelledby="switchLabel" />
    </>
  )
  const toggleSwitch = getByLabelText(SWITCH_LABEL_TEXT)

  expect(toggleSwitch).toHaveAttribute('aria-checked', 'false')
  userEvent.click(toggleSwitch)
  expect(toggleSwitch).toHaveAttribute('aria-checked', 'true')
})
it('switches from off to on with a controlled prop', () => {
  const ControlledSwitchComponent = () => {
    const [isOn, setIsOn] = React.useState(false)

    const onClick = () => {
      setIsOn(!isOn)
    }

    return (
      <>
        <div id="switchLabel">{SWITCH_LABEL_TEXT}</div>
        <Switch onClick={onClick} on={isOn} aria-labelledby="switchLabel" />
      </>
    )
  }
  const {getByLabelText} = render(<ControlledSwitchComponent />)
  const toggleSwitch = getByLabelText(SWITCH_LABEL_TEXT)

  expect(toggleSwitch).toHaveAttribute('aria-checked', 'false')
  userEvent.click(toggleSwitch)
  expect(toggleSwitch).toHaveAttribute('aria-checked', 'true')
})
it('calls onChange when the switch is toggled', () => {
  const handleChange = jest.fn()
  const ControlledSwitchComponent = ({handleSwitchChange}: {handleSwitchChange: (on: boolean) => void}) => {
    const [isOn, setIsOn] = React.useState(false)

    const onClick = () => {
      setIsOn(!isOn)
    }

    return (
      <>
        <div id="switchLabel">{SWITCH_LABEL_TEXT}</div>
        <Switch onClick={onClick} onChange={handleSwitchChange} on={isOn} aria-labelledby="switchLabel" />
      </>
    )
  }
  const {getByLabelText} = render(<ControlledSwitchComponent handleSwitchChange={handleChange} />)
  const toggleSwitch = getByLabelText(SWITCH_LABEL_TEXT)

  userEvent.click(toggleSwitch)
  expect(handleChange).toHaveBeenCalledWith(true)
})

checkStoriesForAxeViolations('Switch/fixtures')
checkStoriesForAxeViolations('Switch/examples')
