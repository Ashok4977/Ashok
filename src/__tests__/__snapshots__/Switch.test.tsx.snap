// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders consistently 1`] = `
.c2 {
  text-align: right;
  visibility: hidden;
  height: 0;
}

.c3 {
  text-align: right;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  overflow: hidden;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  color: #0969da;
  line-height: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 50%;
  -ms-flex-preferred-size: 50%;
  flex-basis: 50%;
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
}

.c7 > svg {
  fill: currentcolor;
}

.c8 {
  color: #24292f;
  line-height: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 50%;
  -ms-flex-preferred-size: 50%;
  flex-basis: 50%;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
}

.c8 > svg {
  fill: currentcolor;
}

.c9 {
  background-color: #f6f8fa;
  width: 50%;
  border-width: 1px;
  border-style: solid;
  border-color: #d0d7de;
  border-radius: 6px;
  position: absolute;
  top: 0;
  bottom: 0;
  z-index: 1;
  box-shadow: rgba(0,0,0,0.12) 0px 2px 1px -1px,rgba(0,0,0,0.1) 0px 1px 4px 0px,rgb(0,0,0,0.08) 0px 1px 3px 0px;
  -webkit-transition-property: -webkit-transform,background-color,border-color;
  -webkit-transition-property: transform,background-color,border-color;
  transition-property: transform,background-color,border-color;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transform: translateX(-1px);
  -ms-transform: translateX(-1px);
  transform: translateX(-1px);
}

.c0 {
  margin-bottom: 24px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  white-space: nowrap;
  border-width: 0;
}

.c1 {
  font-size: 14px;
  color: #24292f;
  margin-left: 8px;
  margin-right: 8px;
  position: relative;
}

.c4 {
  border: none;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 0;
  -webkit-transition-property: background-color,border-color;
  transition-property: background-color,border-color;
  -webkit-transition-delay: 50ms;
  transition-delay: 50ms;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  background: rgba(175,184,193,0.2);
  border-radius: 6px;
  display: block;
  height: 32px;
  width: 64px;
  position: relative;
}

.c4:after {
  content: '';
  box-sizing: border-box;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border-style: solid;
  border-width: 1px;
  border-color: rgba(27,31,36,0.15);
  opacity: 1;
  border-radius: 6px;
}

.c4:hover,
.c4:focus:focus-visible {
  background-color: rgba(27,31,36,0.15);
}

.c4:active:after {
  background-color: rgba(27,31,36,0.15);
  opacity: 0.5;
}

.c4:disabled {
  cursor: not-allowed;
}

<div
  className="c0"
  display="inline-flex"
>
  <span
    aria-hidden="true"
    className="c1"
    color="fg.default"
    fontSize={1}
  >
    <div
      className="c2"
    >
      On
    </div>
    <div
      className="c3"
    >
      Off
    </div>
  </span>
  <button
    aria-checked={false}
    aria-disabled={false}
    className="c4"
    disabled={false}
    on={false}
    onClick={[Function]}
    role="switch"
    size="medium"
  >
    <span
      className="c5"
    >
      Off
    </span>
    <div
      aria-hidden="true"
      className="c6"
      display="flex"
      height="100%"
      overflow="hidden"
      width="100%"
    >
      <div
        className="c7"
        color="accent.fg"
      >
        <svg
          fill="none"
          height="16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M8 2a.75.75 0 0 1 .75.75v11.5a.75.75 0 0 1-1.5 0V2.75A.75.75 0 0 1 8 2Z"
            fillRule="evenodd"
          />
        </svg>
      </div>
      <div
        className="c8"
        color="fg.default"
      >
        <svg
          fill="none"
          height="16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M8 12.5a4.5 4.5 0 1 0 0-9 4.5 4.5 0 0 0 0 9ZM8 14A6 6 0 1 0 8 2a6 6 0 0 0 0 12Z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      aria-hidden="true"
      className="c9"
      width="50%"
    />
  </button>
</div>
`;
